// gitbook-plugin-fullsize-image/index.js
// Main file for the fullsize-image gitbook plugin
'use strict';

module.exports = {
  blocks: {
      image: {
        // Add an "Main Author(s)" block, presumably at the top of a page
        process: function(block) {

          function genAttribs() {
            // Support selected attributes for the image object
            var img_atts = '';

            // width
            var width = block.kwargs['width'];
            if (width) {
              img_atts = 'width="###WIDTH###"'.replace('###WIDTH###', width);
            }

            // height
            var height = block.kwargs['height'];
            if (height) {
              img_atts += 'height="###HEIGHT###"'.replace('###HEIGHT###', height);
            }

            return img_atts;
          }

          function genImage() {
            // If the 'href' kwarg is set
            // return an <img> with surrounding <a>,
            // otherwise the plain image object.
            // TODO: the relative/absolute links to the target don't work yet

            // Define templates
            var image_tpl = '<div class="image"><img src="###IMGSRC###"###IMGATTS### /></div>';
            var href_tpl = '<a href="###HREF###">###IMG###</a>';

            // Dynamically replace placeholders with image URL and attributes
            var img = image_tpl.replace('###IMGSRC###', block.body.trim()).replace('###IMGATTS###', genAttribs());

            // Test if 'href' is present
            var href = block.kwargs['href'];
            if (href) {
              img = href_tpl.replace('###HREF###', href).replace('###IMG###', img);
            };
            return img;
          };

          function genCaption() {
            // If the 'caption' kwarg is set
            // return a <div> with the caption,
            // otherwise an empty string
            var cpt = block.kwargs['caption'];
            if (cpt) {
              var href = block.kwargs['href'];
              if (href) {
                cpt += ' (click to view at full size)'
              }
              return '<div class="caption">###CAPTION###</div>'.replace('###CAPTION###', cpt);
            } else {
              return "";
            }
          };

          // Define templates
          var plugin_tpl = '<div class="image-with-caption">###IMAGEDIV### ###CAPTIONDIV###</div>';

          return plugin_tpl.replace('###IMAGEDIV###', genImage()).replace('###CAPTIONDIV###', genCaption());
        }
      }
  }
};
